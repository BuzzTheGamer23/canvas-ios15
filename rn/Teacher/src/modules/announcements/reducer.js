/* @flow */

import { Reducer } from 'redux'
import { handleActions } from 'redux-actions'
import handleAsync from '../../utils/handleAsync'
import { asyncRefsReducer } from '../../redux/async-refs-reducer'
import { default as ListActions } from './list/actions'
import { default as EditActions } from '../discussions/edit/actions'
import i18n from 'format-message'
import composeReducers from '../../redux/compose-reducers'

const { refreshAnnouncements } = ListActions
const { createDiscussion, deleteDiscussion } = EditActions

const list: Reducer<AsyncRefs, any> = asyncRefsReducer(
  refreshAnnouncements.toString(),
  i18n('There was a problem loading the announcements.'),
  ({ result }) => result.data.map(announcement => announcement.id)
)

const refsChanges: Reducer<AsyncRefs, any> = handleActions({
  [createDiscussion.toString()]: handleAsync({
    resolved: (state, { result: { data } }) => ({
      ...state,
      refs: [...state.refs, data.id],
    }),
  }),
  [deleteDiscussion.toString()]: handleAsync({
    resolved: (state, { discussionID }) => ({
      ...state,
      refs: (state.refs || []).filter(ref => ref !== discussionID),
    }),
  }),
}, {})

export const refs: Reducer<AsyncRefs, any> = composeReducers(list, refsChanges)
