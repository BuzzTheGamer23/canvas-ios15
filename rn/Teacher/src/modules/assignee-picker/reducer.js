// @flow

import { Reducer } from 'redux'
import SectionActions from './actions'
import { handleActions } from 'redux-actions'
import handleAsync from '../../utils/handleAsync'
import fromPairs from 'lodash/fromPairs'

const { refreshSections } = SectionActions

export const sections: Reducer<any, any> = handleActions({
  [refreshSections.toString()]: handleAsync({
    resolved: (state, { result }) => {
      const incoming = fromPairs(result.data
        .map(section => [section.id, section]))
      return { ...state, ...incoming }
    },
  }),
}, {})
