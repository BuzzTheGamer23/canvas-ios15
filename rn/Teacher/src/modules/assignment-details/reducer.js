/* @flow */

import { Reducer } from 'redux'
import { handleActions } from 'redux-actions'
import Actions from './actions'
import type { AssignmentDetailsState } from './props'
import handleAsync from '../../utils/handleAsync'
import i18n from 'format-message'

export let defaultState: { assignmentDetails: any, pending: number, nextPage: ?Function } = { assignmentDetails: {}, pending: 0, nextPage: null }

const { refreshAssignmentDetails } = Actions

const reducer: Reducer<AssignmentDetailsState, any> = handleActions({
  [refreshAssignmentDetails.toString()]: handleAsync({
    pending: (state) => ({ ...state, pending: state.pending + 1 }),
    resolved: (state, response) => {
      return {
        ...state,
        pending: state.pending - 1,
        assignmentDetails: response.data,
      }
    },
    rejected: (state, response) => {
      let errorMessage = i18n('Could not get assignment list')
      return {
        ...state,
        error: errorMessage,
        pending: state.pending - 1,
      }
    },
  }),
}, defaultState)

export default reducer
