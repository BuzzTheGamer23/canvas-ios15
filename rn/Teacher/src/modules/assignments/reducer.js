/* @flow */

import { Reducer } from 'redux'
import { handleActions } from 'redux-actions'
import Actions from './actions'
import type { AssignmentListState } from './props'
import handleAsync from '../../utils/handleAsync'
import i18n from 'format-message'

export let defaultState: { assignmentGroups: AssignmentGroup[], pending: number, nextPage: ?Function } = { assignmentGroups: [], pending: 0, nextPage: null }

const { refreshAssignmentList } = Actions

const reducer: Reducer<AssignmentListState, any> = handleActions({
  [refreshAssignmentList.toString()]: handleAsync({
    pending: (state) => ({ ...state, pending: state.pending + 1 }),
    resolved: (state, paginatedResponse) => {
      const assignmentGroups = [...state.assignmentGroups, ...paginatedResponse.data].sort((a, b) => a.position - b.position)
      return {
        ...state,
        pending: state.pending - 1,
        assignmentGroups,
        nextPage: paginatedResponse.next,
      }
    },
    rejected: (state, response) => {
      let errorMessage = i18n('Could not get assignment list')
      return {
        ...state,
        error: errorMessage,
        pending: state.pending - 1,
      }
    },
  }),
}, defaultState)

export default reducer
