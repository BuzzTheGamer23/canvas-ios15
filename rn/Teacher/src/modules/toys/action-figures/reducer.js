// @flow

// import * as actions from './actions'

import { Action } from 'redux'

export default function actionFigures (state: string[] = [], action: Action): [] {
  switch (action.type) {
    default:
      return state
  }
}
