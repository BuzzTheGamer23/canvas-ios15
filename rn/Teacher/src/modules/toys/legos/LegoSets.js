// @flow

import React, { Component, PropTypes } from 'react'
import {
  ActivityIndicator,
  View,
  Text,
  Image,
  Button,
} from 'react-native'
import { connect } from 'react-redux'
import { stateToProps } from './props'
import type { LegoSetsProps } from './props'
import LegoActions from './actions'
import i18n from 'format-message'

export class DisconnectedLegoSets extends Component<any, LegoSetsProps, any> {

  renderLegoSet = (set: LegoSet, index: number) => {
    return (
      <View key={index} testID={set.imageURL}>
        <Image source={{ uri: set.imageURL }} />
        <Text>{ set.name }</Text>
      </View>
    )
  }

  buyTheFalcon = () => {
    this.props.buyLegoSet({
      name: 'The Millenium Falcon',
      imageURL: 'https://lego.com/falcon',
    })
  }

  render (): React.Element<{}> {
    const sets = this.props.sets.map(this.renderLegoSet)
    const activity = this.props.pending > 0
    const indicator = activity ? <ActivityIndicator /> : <View />

    return (
      <View>
        <Text>Lego Sets!</Text>
        {sets}
        {indicator}
        <Text style={{ color: 'red' }}>{this.props.error}</Text>
        <Button
          testID={'toys.legos.buyLegos'}
          title={ i18n('Buy More!') }
          onPress={this.buyTheFalcon} />
        <Button
          title={ i18n('Out of money, sell them all!') }
          onPress={this.props.sellAllLegos} />
      </View>
    )
  }
}

const legoSetShape = PropTypes.shape({
  name: PropTypes.string.isRequired,
  imageURL: PropTypes.string.isRequired,
}).isRequired

DisconnectedLegoSets.propTypes = {
  sets: PropTypes.arrayOf(legoSetShape).isRequired,
  pending: PropTypes.number,
  error: PropTypes.string,
}

export default connect(stateToProps, LegoActions)(DisconnectedLegoSets)
