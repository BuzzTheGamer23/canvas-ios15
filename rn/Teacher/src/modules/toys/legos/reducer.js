// @flow

import { Reducer } from 'redux'
import { handleActions } from 'redux-actions'
import LegoActions from './actions'
import type { LegoState } from './props'
import handleAsync from '../../../utils/handleAsync'

export let defaultState = { sets: [], pending: 0 }

let { buyLegoSet, sellAllLegos } = LegoActions

const reducer: Reducer<LegoState, any> = handleActions({
  [buyLegoSet.toString()]: handleAsync({
    pending: (state) => ({ ...state, pending: state.pending + 1 }),
    resolved: (state, response) => ({
      ...state,
      sets: state.sets.concat(response.data),
      pending: state.pending - 1,
    }),
    rejected: (state, error) => ({
      ...state,
      error: 'no monies ;(',
      pending: state.pending - 1,
    }),
  }),
  [sellAllLegos.toString()]: handleAsync({
    pending: (state) => ({ ...state, pending: state.pending + 1 }),
    resolved: (state) => ({ ...state, sets: [], pending: state.pending - 1 }),
  }),
}, defaultState)

export default reducer
