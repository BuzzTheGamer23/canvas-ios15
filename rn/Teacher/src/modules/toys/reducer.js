// @flow

import { combineReducers } from 'redux'
import type { Reducer } from 'redux'

import legoSets from './legos/reducer'
import type { LegoState } from './legos/props'
import actionFigures from './action-figures/reducer'

export type ToyState = {
  legoSets: LegoState,
  actionFigures: any,
}

export default (combineReducers({
  legoSets,
  actionFigures,
}): Reducer<ToyState, any>)
