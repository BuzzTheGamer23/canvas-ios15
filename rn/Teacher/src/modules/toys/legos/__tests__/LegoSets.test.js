// @flow

const { it, expect } = global
import { DisconnectedLegoSets } from '../LegoSets'
import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import explore from '../../../../../test/helpers/explore'

import { defaultState } from '../reducer'

declare var jest: any

jest.mock('Button', () => 'Button')

it('renders 0 sets correctly', () => {
  const props = defaultState
  const tree = renderer.create(
    <DisconnectedLegoSets { ...props } />
  ).toJSON()

  expect(tree).toMatchSnapshot('0-lego-sets')
})

it('renders 2 sets correctly', () => {
  const sets = [
    {
      name: 'The Batcave!',
      imageURL: 'https://legos.com/batcave',
    },
    {
      name: 'Epic Pirate Ship',
      imageURL: 'https://lego.com/pirates',
    },
  ]
  const props = { sets }
  const tree = renderer.create(
    <DisconnectedLegoSets { ...props } />
  ).toJSON()

  expect(tree).toMatchSnapshot('2-lego-sets')
})

it('can buy the Millenium Falcon set!', () => {
  const buyLegoSet = jest.fn()

  const props = {
    sets: [],
    buyLegoSet,
  }
  const tree = renderer.create(
    <DisconnectedLegoSets { ...props } />
  ).toJSON()

  const button: any = explore(tree).selectByID('toys.legos.buyLegos')
  button.props.onPress()

  expect(buyLegoSet).toHaveBeenCalledWith({
    name: 'The Millenium Falcon',
    imageURL: 'https://lego.com/falcon',
  })
})
