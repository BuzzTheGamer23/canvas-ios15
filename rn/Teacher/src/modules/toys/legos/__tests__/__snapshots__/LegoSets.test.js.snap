// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0-lego-sets 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Lego Sets!
  </Text>
  <View />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "red",
      }
    }
  />
  <Button
    onPress={[Function]}
    testID="toys.legos.buyLegos"
    title="Buy More!"
  />
  <Button
    onPress={undefined}
    title="Out of money, sell them all!"
  />
</View>
`;

exports[`2-lego-sets 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Lego Sets!
  </Text>
  <View
    testID="https://legos.com/batcave"
  >
    <Image
      source={
        Object {
          "uri": "https://legos.com/batcave",
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      The Batcave!
    </Text>
  </View>
  <View
    testID="https://lego.com/pirates"
  >
    <Image
      source={
        Object {
          "uri": "https://lego.com/pirates",
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Epic Pirate Ship
    </Text>
  </View>
  <View />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "red",
      }
    }
  />
  <Button
    onPress={[Function]}
    testID="toys.legos.buyLegos"
    title="Buy More!"
  />
  <Button
    onPress={undefined}
    title="Out of money, sell them all!"
  />
</View>
`;
